### PR TITLE
Add truffle_managed_malloc

### PR DESCRIPTION
**Pull request by chrisseaton**

### CODE REVIEW

**Comment by mrigger** [comment]
Could you also shortly document the function in `INTEROP.md`? I think we should do that from now on for every change in the interop API.

**Comment by chrisseaton** [comment]
Sure - just doing this to run the tests!

File: projects/com.oracle.truffle.llvm.nodes.impl/src/com/oracle/truffle/llvm/nodes/impl/memory/load/LLVMDirectLoadNode.java [hunk 1]
```diff
@@ -91,6 +93,22 @@ public LLVMFunctionDescriptor executeAddress(LLVMAddress addr) {
...
         }
+
+        // The following specializations are simpler than supporting arbitrary foreign objects here
+
+        @Specialization
+        public Object executeAddress(ManagedMallocObject addr) {
+            return addr.get(0);
+        }
+
+        @Specialization(guards = "objectIsManagedMalloc(addr)")
```
Comment (mrigger) [inline projects/com.oracle.truffle.llvm.nodes.impl/src/com/oracle/truffle/llvm/nodes/impl/memory/load/LLVMDirectLoadNode.java]: Can this specialization occur? Isn't it already covered by the first specialization?

File: projects/com.oracle.truffle.llvm.nodes.impl/src/com/oracle/truffle/llvm/nodes/impl/memory/load/LLVMDirectLoadNode.java [hunk 2]
```diff
@@ -91,6 +93,22 @@ public LLVMFunctionDescriptor executeAddress(LLVMAddress addr) {
         public LLVMAddress executeAddress(LLVMAddress addr) {
             return LLVMMemory.getAddress(addr);
         }
+
```
Comment (mrigger) [inline projects/com.oracle.truffle.llvm.nodes.impl/src/com/oracle/truffle/llvm/nodes/impl/memory/load/LLVMDirectLoadNode.java]: Can we change the `@NodeChild` to `LLVMExpressionNode`?

File: projects/com.oracle.truffle.llvm.nodes.impl/src/com/oracle/truffle/llvm/nodes/impl/memory/load/LLVMDirectLoadNode.java (hunk 1 above)
Comment (chrisseaton) [inline projects/com.oracle.truffle.llvm.nodes.impl/src/com/oracle/truffle/llvm/nodes/impl/memory/load/LLVMDirectLoadNode.java]: Yes the first is the case of dereferencing a managed malloc with either `*value` or `value[0]` when the `0` is constant, and the second is the case of an arbitrary index or an inner pointer.

File: projects/com.oracle.truffle.llvm.nodes.impl/src/com/oracle/truffle/llvm/nodes/impl/memory/load/LLVMDirectLoadNode.java (hunk 2 above)
Comment (chrisseaton) [inline projects/com.oracle.truffle.llvm.nodes.impl/src/com/oracle/truffle/llvm/nodes/impl/memory/load/LLVMDirectLoadNode.java]: The `@NodeChild` where? There isn't one in this class is there?

File: projects/com.oracle.truffle.llvm.nodes.impl/src/com/oracle/truffle/llvm/nodes/impl/memory/load/LLVMDirectLoadNode.java (hunk 1 above)
Comment (mrigger) [inline projects/com.oracle.truffle.llvm.nodes.impl/src/com/oracle/truffle/llvm/nodes/impl/memory/load/LLVMDirectLoadNode.java]: Ah, sorry. I overlooked that.

File: projects/com.oracle.truffle.llvm.nodes.impl/src/com/oracle/truffle/llvm/nodes/impl/memory/load/LLVMDirectLoadNode.java (hunk 2 above)
Comment (mrigger) [inline projects/com.oracle.truffle.llvm.nodes.impl/src/com/oracle/truffle/llvm/nodes/impl/memory/load/LLVMDirectLoadNode.java]: here: https://github.com/graalvm/sulong/blob/a30dd33109b07190bd974dc03c0d7d2d8db22259/projects/com.oracle.truffle.llvm.nodes.impl/src/com/oracle/truffle/llvm/nodes/impl/memory/load/LLVMDirectLoadNode.java#L89

Comment (chrisseaton) [inline projects/com.oracle.truffle.llvm.nodes.impl/src/com/oracle/truffle/llvm/nodes/impl/memory/load/LLVMDirectLoadNode.java]: Done.

**Pull request merged**